### PR TITLE
Fixed compiler warnings

### PR DESCRIPTION
**Pull request by larsewi**
- **fix: compiler warning -Werror=unused-function**
- **fix: compiler warning -Werror=sign-compare**
- **fix: compiler warning -Werror=pointer-sign**
- **fix: compiler warning -Werror=pointer-arith**
- **fix: compiler warning -Werror=implicit-function-declaration**
- **build: added option to enable all warnings as errors**
- **test: enabled all warnings as errors in smoke test**

### CODE REVIEW

File: core/src/mender-artifact.c [hunk 1]
```diff
@@ -552,7 +552,8 @@ mender_artifact_read_manifest(mender_artifact_ctx_t *ctx) {
     return MENDER_DONE;
 }
 
-static mender_err_t
+/* TODO: Should we remove unused function? */
+__attribute__((unused)) static mender_err_t
 mender_create_provides_depends_node(const char *type, const char *value, mender_key_value_list_t **provides_depends) {
```
Comment (larsewi) [inline core/src/mender-artifact.c:557]: @danielskinstad do you know if we plan to use this function, or should we remove it?

Comment (danielskinstad) [reply core/src/mender-artifact.c:557]: We should remove it. I moved it to utils as `mender_utils_create_key_value_node`, but apparently forgot to remove it here.

File: platform/tls/generic/mbedtls/src/mender-tls.c [hunk 2]
```diff
@@ -447,7 +447,7 @@ mender_tls_user_provided_authentication_keys(mbedtls_pk_context *pk_context, con
 #if MBEDTLS_VERSION_NUMBER >= 0x03000000
     if (0 != (ret = mbedtls_pk_parse_key(pk_context, user_provided_key, user_provided_key_length, NULL, 0, mbedtls_ctr_drbg_random, ctr_drbg))) {
 #else
-    if (0 != (ret = mbedtls_pk_parse_key(pk_context, user_provided_key, user_provided_key_length, NULL, 0))) {
+    if (0 != (ret = mbedtls_pk_parse_key(pk_context, (const unsigned char *)user_provided_key, user_provided_key_length, NULL, 0))) {
```
Comment (lluiscampos) [inline platform/tls/generic/mbedtls/src/mender-tls.c:450]: Please fix this also above. Latest Zephy LTS release comes with the most recent mbedtls version

**Review by lluiscampos — APPROVED** [review]

**Comment by lluiscampos** [comment]
Awesome. Thank you

File: core/src/mender-artifact.c [hunk 3]
```diff
@@ -505,7 +505,7 @@ mender_artifact_read_manifest(mender_artifact_ctx_t *ctx) {
 
     /* Read data line by line */
     char *line = ctx->input.data;
-    char *end  = ctx->input.data + ctx->input.length;
+    char *end  = (char *)ctx->input.data + ctx->input.length;
```
Comment (vpodzime) [inline core/src/mender-artifact.c:508]: Why not use `line` instead?

File: CMakeLists.txt [hunk 4]
```diff
@@ -21,6 +21,11 @@ cmake_minimum_required(VERSION 3.16.3)
 # Library project
 project(mender-mcu-client LANGUAGES C)
 
+# Compiler options
+if (CONFIG_MENDER_ALL_WARNINGS_AS_ERRORS)
+    add_compile_options(-Wall -Wextra -Werror -Wpointer-arith)
```
Comment (vpodzime) [inline CMakeLists.txt:26]: So `-Wpointer-arith` is not part of neither `-Wall`  nor `-Wextra`? That's really strange.

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -46,5 +46,6 @@ test:smoke-build:
       -DCONFIG_MENDER_CLIENT_ADD_ON_TROUBLESHOOT=OFF
       -DCONFIG_MENDER_FULL_PARSE_ARTIFACT=ON
       -DCONFIG_MENDER_PROVIDES_DEPENDS=ON
+      -DCONFIG_MENDER_ALL_WARNINGS_AS_ERRORS=ON
```
Comment (vpodzime) [inline .gitlab-ci.yml:49]: Do we have the script to run the above yet? It should be in some place with a proper name and available for people to run this when testing things locally.

Comment (larsewi) [reply .gitlab-ci.yml:49]: This think @danielskinstad is working on that here https://github.com/mendersoftware/mender-mcu/pull/30

File: core/src/mender-artifact.c (hunk 3 above)
Comment (larsewi) [reply core/src/mender-artifact.c:508]: Yes, I will fix 🚀

**Pull request merged**